debezium-server-kinesis/src/main/java/io/debezium/server/kinesis/KinesisChangeConsumer.java

```diff
@@ -106,41 +125,102 @@ void close() {
+                batch = segmentedBatch.subList(i, Math.min(i + batchSize, segmentedBatch.size()));
+                List<PutRecordsRequestEntry> putRecordsRequestEntryList = new ArrayList<>();
+                streamName = batch.get(0).destination();
+
+                for (ChangeEvent<Object, Object> record : batch) {
+
+                    Object rv = record.value();
+                    if (rv == null) {
+                        rv = "";
+                    }
+                    PutRecordsRequestEntry putRecordsRequestEntry = PutRecordsRequestEntry.builder()
+                            .partitionKey((record.key() != null) ? getString(record.key()) : nullKey)
+                            .data(SdkBytes.fromByteArray(getBytes(rv))).build();
+                    putRecordsRequestEntryList.add(putRecordsRequestEntry);
+                }
+
+                // Handle Error
+                boolean notSuccesful = true;
+                int attempts = 0;
+                List<PutRecordsRequestEntry> batchRequest = putRecordsRequestEntryList;
+
+                while (notSuccesful) {
+
+                    if (attempts >= RETRIES) {
+                        throw new DebeziumException("Exceeded maximum number of attempts to publish event");
+                    }
+
+                    try {
+                        PutRecordsResponse response = recordsSent(batchRequest, streamName);
+                        attempts++;
+                        if (response.failedRecordCount() > 0) {
+                            LOGGER.warn("Failed to send {} number of records, retrying", response.failedRecordCount());
+                            Metronome.sleeper(RETRY_INTERVAL, Clock.SYSTEM).pause();
+
+                            final List<PutRecordsResultEntry> putRecordsResults = response.records();
+                            List<PutRecordsRequestEntry> failedRecordsList = new ArrayList<>();
+
+                            for (int index = 0; index < putRecordsResults.size(); index++) {
+                                PutRecordsResultEntry entryResult = putRecordsResults.get(index);
+                                if (entryResult.errorCode() != null) {
+                                    failedRecordsList.add(putRecordsRequestEntryList.get(index));
+                                }
+                            }
+                            batchRequest = failedRecordsList;
+
+                        }
+                        else {
+                            notSuccesful = false;
+                        }
+
+                    }
+                    catch (KinesisException exception) {
+                        LOGGER.warn("Failed to send record to {}", streamName, exception);
+                        attempts++;
+                        Metronome.sleeper(RETRY_INTERVAL, Clock.SYSTEM).pause();
+                    }
+                }
+
+                for (ChangeEvent<Object, Object> record : batch) {
```

Comment:
As the batches are grouped by topic they will not be sent in total order - this is expected. Do you want also mark records as processed per-batch even if it does not conform to the total order or would you prefer to move it out of the batch loop and mark them at the end in a single loop over original `records` list?